### PR TITLE
Add `mypy` to the pipeline

### PR DESCRIPTION
**Pull request by Kludex**
- Replaces #7325
- Related to #7258

Thanks @atombrella for the initial work!

This PR adds `mypy` to `tox` and to the pipeline. A couple of differences between this PR and the previous one:
- Apply mypy per file. Meaning that we can do progressive work just adding a new file to the `pyproject.toml`.
- Use `pyproject.toml` instead of `mypy.ini`.
- `disallow_untyped_defs` is `True` now.

This PR is intended to be minimal, so we can start working _per files basis_. I've reused the previous branch, so there are also some small type annotation fixes, which are not required for this PR.

### CODE REVIEW

**Comment by povilasb** [comment]
Nice, type hints are much appreciated when navigating the code base.
Thank you!

**Comment by Kludex** [comment]
Yep. As soon as this gets in, I'll organize a similar issue as I did on `uvicorn`: https://github.com/encode/uvicorn/issues/998

Then we can start working on it :)

File: celery/contrib/testing/worker.py [hunk 1]
```diff
@@ -90,16 +90,15 @@ def start_worker(
...
+                         pool: str = 'solo',
                          loglevel=WORKER_LOGLEVEL,
-                         logfile=None,
-                         WorkController=TestWorkController,
-                         perform_ping_check=True,
-                         shutdown_timeout=10.0,
-                         **kwargs):
-    # type: (Celery, int, str, Union[str, int], str, Any, **Any) -> Iterable
+                         logfile: Union[str, int] = None,
+                         WorkController: Any = TestWorkController,
```
Comment (auvipy) [inline celery/contrib/testing/worker.py:98]: shouldn't we try not to use Any?

**Review by auvipy — COMMENTED** [review]
good job

File: celery/contrib/testing/worker.py (hunk 1 above)
Comment (Kludex) [reply celery/contrib/testing/worker.py:98]: Yes. I didn't remove the annotations added from the previous PR. Let me check them.

**Comment by Kludex** [comment]
All the annotations besides `__main__.py` are not needed for this PR. If you want, I can remove the other ones.

File: celery/utils/saferepr.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
 from itertools import chain
 from numbers import Number
 from pprint import _recursion
-from typing import Any, AnyStr, Callable, Dict, Iterator, List, Sequence, Set, Tuple
+from typing import Any, AnyStr, Callable, Dict, Iterator, List, Optional, Sequence, Set, Tuple
```
Comment (Kludex) [inline celery/utils/saferepr.py:18]:
```suggestion
from typing import Any, AnyStr, Callable, Dict, Iterator, List, Sequence, Set, Tuple
```

**Comment by Kludex** [comment]
I've just found out about this: https://github.com/encode/uvicorn/pull/1437#issuecomment-1086574265

Might be useful for celery as well.

**Review by auvipy — APPROVED** [review]

**Comment by Kludex** [comment]
Thanks @auvipy ! :)

I'm going to create an issue so we can organize the work towards a fully annotated celery! 🙌

**Comment by auvipy** [comment]
we can just modify this https://github.com/celery/celery/issues/7258

**Comment by Kludex** [comment]
@auvipy I didn't modify it because I'd like to have rights to edit. See the created issue: #7394

**Pull request merged**